src/reflect/scala/reflect/internal/Names.scala

```diff
@@ -41,16 +41,19 @@ trait Names extends api.Names {
   private val typeHashtable = new Array[TypeName](HASH_SIZE)
 
   /**
-   * The hashcode of a name depends on the first, the last and the middle character,
+   * The hashcode of is equivalent to cs name depends on the first, the last and the middle character,
    * and the length of the name.
    */
-  private def hashValue(cs: Array[Char], offset: Int, len: Int): Int =
-    if (len > 0)
-      (len * (41 * 41 * 41) +
-       cs(offset) * (41 * 41) +
-       cs(offset + len - 1) * 41 +
-       cs(offset + (len >> 1)))
-    else 0
+  private def hashValue(cs: Array[Char], offset: Int, len: Int): Int = {
```

Comment:
@lrytz, 
>@DarkDimius how do you guarantee that strings don't repeat? Let's say you first create a name "abc", then a name "bc", will the name table only contain "abc"?

It would contain both, but the next time you try to create `abc` you'll get the same one. See https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/core/Names.scala#L245

> Also, to me, dotty's Names.scala looks quite similar to scala's. There's a hashValue method that looks the same as the one being replaced in this PR. Can you point out differences you have in mind?

hashValue only used when creating new Term names. It's not used when comparing hashcodes.
Hashcode is https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/core/Names.scala#L178